foundations/javascript_basics/understanding_errors.md

```diff
@@ -0,0 +1,132 @@
+
+
+We’ll learn more about the different error types later in the lesson. Error types denote the kind of error, which gives us clues about how to address them, and why they may arise.
+
+In this case, a `ReferenceError` is an error that is thrown when one attempts to reference a variable that has not yet been declared. The error message also explains this, by stating that `c is not defined`. Different kinds of errors can have different messages based on what is actually causing the `ReferenceError`. For example, another message you may run into that is also a `ReferenceError` is `ReferenceError: can't access lexical declaration 'X' before initialization`. As we can see, this points to a completely different reason than our original `ReferenceError` above. Understanding both the error type and the error message is crucial to comprehending why you are receiving the error.
+
+The next part of an error gives us the line of code, as well as the file in which you can find the error (in this case, our `script.js`). This allows you to easily navigate to the problematic line in your code. Here, the error originates from the fifth line of `script.js`, which is displayed as a link under the error message with the text `at script.js:5`. If you click this link, most browsers will navigate to the exact line of code and the rest of your script in the Sources tab of the Developer Tools. 
+
+Another important part of an error is the stack trace. While the line number helps you locate the line in your code, the stack trace helps you understand when the error was thrown in your application, and what functions were called that lead up to running the lines of code responsible for the error. So for example, if we have the following code:
+
+```js
+const a = 5;
+const b = 10;
+
+function add() {
+  return c;
+}
+
+function print() {
+  add();
+}
+
+print();
+```
+
+Our function `print()` should call on `add()`, which returns a variable named `c`, which currently has not been declared. The corresponding error is as follows:
+
+<img width="620" alt="Screen Shot 2021-05-17 at 12 15 06 PM" src="https://user-images.githubusercontent.com/22967723/118543770-90641980-b709-11eb-8c85-4f49195901bc.png">
+
+The stack trace tells us that:
+
+1. `c is not defined` in scope of `add()`, which is declared on line 5
+2. `add()` was called by `print()`, which was declared on line 9
+3. `print()` itself was called on line 12.
+   The stack trace allows you to follow the execution of functions to find the code responsible for throwing the error .
+
+# Common Types of Errors
+
+These are some of the most common errors you will encounter, so it’s important to understand them.
+
+## Syntax Error
+
+A syntax error occurs when the code you are trying to run is syntactically incorrect. For example this:
+
+```js
+function helloWorld() {
+console.log “Hello World!”
```

Comment:
Missing indentation. I don't think we want to imply that lack of indentation can cause an error.